### PR TITLE
nginx: some http3 module fixes

### PR DESCRIPTION
**Pull request by mkg20001**
###### Motivation for this change

Make it better

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by SuperSandro2000** [comment]
Why did you disable reuseport? Does it cause issues?

**Comment by mkg20001** [comment]
Yes. It's only possible to have one virtual server listening with reuseport. Creating multiple fails. https://trac.nginx.org/nginx/ticket/2179

**Comment by jtojnar** [comment]
Could you mention that in the commit message?

**Comment by mkg20001** [comment]
Added it

File: nixos/modules/services/web-servers/nginx/default.nix [hunk 1]
```diff
@@ -254,7 +254,7 @@ let
           # UDP listener for **QUIC+HTTP/3
           listen ${addr}:${toString port} http3 reuseport;
           # Advertise that HTTP/3 is available
-          add_header Alt-Svc 'h3=":443"';
+          add_header Alt-Svc 'h3-29=":443"; ma=86400';
```
Comment (mweinelt) [inline nixos/modules/services/web-servers/nginx/default.nix:257]: You should probably explain what is fixed by this and how.

**Comment by ajs124** [comment]
I'm not really in favor of just disabling reuseport. AFAIR, you can only specify it once in the config, but that means it will just be enabled. So basically, it just needs to be specified in the first http3 listening directive.

**Comment by Izorkin** [comment]
@mkg20001 @ajs124 
Option `reuseport` don`t needed. Because of this, the HTTP3 protocol does not work with multiple hosts.

Minimal working configuration without `reuseport`:
```
pid /run/nginx/nginx.pid;
error_log stderr;
daemon off;
events { }
http {
  include /nix/store/qd3g8rk5hx5zkb70idjh6fa12sh6bipg-mailcap-2.1.53/etc/nginx/mime.types;
  include /nix/store/mwibmbkyhimr43bkbrpz9sway6krcniq-nginx-quic/conf/fastcgi.conf;
  include /nix/store/mwibmbkyhimr43bkbrpz9sway6krcniq-nginx-quic/conf/uwsgi_params;
  default_type application/octet-stream;
  ssl_protocols TLSv1.2 TLSv1.3;
  ssl_ciphers ECDHE-ECDSA-AES128-GCM-SHA256:ECDHE-RSA-AES128-GCM-SHA256:ECDHE-ECDSA-AES256-GCM-SHA384:ECDHE-RSA-AES256-GCM-SHA384:ECDHE-ECDSA-CHACHA20-POLY1305:ECDHE-RSA-CHACHA20-POLY1305:DHE-RSA-AES128-G>
  client_max_body_size 10m;
  server_tokens off;

  server {
    listen 0.0.0.0:443 http3;
    listen 0.0.0.0:443 ssl http2;
    server_name test1.local;
    root /var/www/test1;
    ssl_certificate /var/certs/test1.crt;
    ssl_certificate_key /var/certs/test1.key;
  }

  server {
    listen 0.0.0.0:443 http3;
    listen 0.0.0.0:443 ssl http2;
    server_name test2.local;
    root /var/www/test2;
    ssl_certificate /var/certs/test2.crt;
    ssl_certificate_key /var/certs/test2.key;
  }
}
```

Also in the current version of the module, duplicate parameters are created:
```
  server {
    listen 0.0.0.0:443 ssl http2 ;# UDP listener for **QUIC+HTTP/3
    listen 0.0.0.0:443 http3 reuseport;
    # Advertise that HTTP/3 is available
    add_header Alt-Svc 'h3=":443"';
    # Sent when QUIC was used
    add_header QUIC-Status $quic;
    listen [::0]:443 ssl http2 ;# UDP listener for **QUIC+HTTP/3
    listen [::0]:443 http3 reuseport;
    # Advertise that HTTP/3 is available
    add_header Alt-Svc 'h3=":443"';
    # Sent when QUIC was used
    add_header QUIC-Status $quic;
    server_name test1.local ;
```

**Comment by Izorkin** [comment]
Optimized patch:
```diff
diff --git a/nixos/modules/services/web-servers/nginx/default.nix b/nixos/modules/services/web-servers/nginx/default.nix
index 3c5ae36652d..8d3cf043297 100644
--- a/nixos/modules/services/web-servers/nginx/default.nix
+++ b/nixos/modules/services/web-servers/nginx/default.nix
@@ -254,20 +254,19 @@ let
             else defaultListen;

         listenString = { addr, port, ssl, extraParameters ? [], ... }:
-          "listen ${addr}:${toString port} "
-          + optionalString ssl "ssl "
+          (if ssl && vhost.http3 then "
+          # UDP listener for **QUIC+HTTP/3
+          listen ${addr}:${toString port} http3 "
+          + optionalString vhost.default "default_server "
+          + ";" else "")
+          + "
+
+            listen ${addr}:${toString port} "
           + optionalString (ssl && vhost.http2) "http2 "
+          + optionalString ssl "ssl "
           + optionalString vhost.default "default_server "
           + optionalString (extraParameters != []) (concatStringsSep " " extraParameters)
-          + ";"
-          + (if ssl && vhost.http3 then ''
-          # UDP listener for **QUIC+HTTP/3
-          listen ${addr}:${toString port} http3 reuseport;
-          # Advertise that HTTP/3 is available
-          add_header Alt-Svc 'h3=":443"';
-          # Sent when QUIC was used
-          add_header QUIC-Status $quic;
-          '' else "");
+          + ";";

         redirectListen = filter (x: !x.ssl) defaultListen;

@@ -317,6 +316,13 @@ let
             ssl_reject_handshake on;
           ''}

+          ${optionalString (hasSSL && vhost.http3) ''
+            # Advertise that HTTP/3 is available
+            add_header Alt-Svc 'h3=":443"; ma=86400';
+            # Sent when QUIC was used
+            add_header QUIC-Status $quic;
+          ''}
+
           ${mkBasicAuth vhostName vhost}

           ${mkLocations vhost.locations}
```

On one of hosts, you need to necessarily add the parameter `reuseport`. Otherwise, it stops working normally.
```
 curl --http3 --head -4 https://example_site
curl: (56) Failure when receiving data from the peer
```

File: nixos/modules/services/web-servers/nginx/default.nix (hunk 1 above)
Comment (dasJ) [reply nixos/modules/services/web-servers/nginx/default.nix:257]: Also why? Firefox, Chrome and curl all work with the h3 ALPN. Also 86400 is 24h which is the default

**Comment by SuperSandro2000** [comment]
can you create a PR with that? Otherwise I would close this.

**Comment by Izorkin** [comment]
@SuperSandro2000 already implemented in this PR - https://github.com/NixOS/nixpkgs/pull/146917

**Pull request closed without merging**